### PR TITLE
Feature/768/payment, closes #768, closes #769, closes #795

### LINKED ISSUE #768: Payment
* http://dev.sphere.io/http-api-projects-payments.html
* http://dev.sphere.io/http-api-projects-orders.html#payment-info
* http://dev.sphere.io/http-api-projects-orders.html#order payment info field

subtasks:

* [x] Payment
* [x] PaymentMethodInfo
* [x] PaymentStatus
* [x] Transaction
* [x] Payment Draft + builder
* [x] Get Payment by ID
* [x] Get Payment by ID test
* [x] Query Payments
* [x] Query Payments test
* [x] Create a Payment
* [x] Create a Payment test
* [x] Update a Payment (only the command)
* [x] setCustomer
* [x] setCustomer test
* [x] setExternalId
* [x] setExternalId test
* [x] setInterfaceId test
* [x] setAuthorization
* [x] setAuthorization test
* [x] setAmountPaid
* [x] setAmountPaid test
* [x] setAmountRefunded
* [x] setAmountRefunded test
* [x] setMethodInfoInterface
* [x] setMethodInfoInterface test
* [x] setMethodInfoMethod
* [x] setMethodInfoMethod test
* [x] setMethodInfoName
* [x] setMethodInfoName test
* [x] setCustomField
* [x] setCustomField test
* [x] setCustomType
* [x] setCustomType test
* [x] setStatusInterfaceCode
* [x] setStatusInterfaceCode test
* [x] setStatusInterfaceText
* [x] setStatusInterfaceText test
* [x] transitionState
* [x] transitionState test
* [x] addTransaction test
* [x] Delete a Payment
* [x] Delete a Payment test
* [ ] check MVP requirements from Payments HipChat room
* [x] report enumeration not sphere ideomatic
* [x] update actions cart
* [x] update actions order

### LINKED ISSUE #769: Payment meta models like query model and expansion model
https://github.com/sphereio/sphere-jvm-sdk/issues/768

* [x] customer
* [x] paymentStatus.state

### LINKED ISSUE #795: NullPointerException with Message.getResource()
see https://github.com/sphereio/sphere-jvm-sdk/pull/786#issuecomment-148209740

### PR DESCRIPTION
**Pull request by schleichardt**

### CODE REVIEW

**Comment by schleichardt** [comment]
@cneijenhuis @lauraluiz The build is green and the Javadoc deployed.

index: http://sphereio.github.io/sphere-jvm-sdk/javadoc/feature/768/payment/index.html
Payment: http://sphereio.github.io/sphere-jvm-sdk/javadoc/feature/768/payment/io/sphere/sdk/payments/Payment.html

**Comment by cneijenhuis** [comment]
`setCustomer`, `setExternalId` and `setAuthorization` update actions are missing (at least in the JavaDoc).

**Comment by schleichardt** [comment]
Yes, that is weird, because they are in the code: https://github.com/sphereio/sphere-jvm-sdk/pull/786/files#diff-1607d5ee0c82b8a84684100017fdbc6aR21

File: sphere-models/src/it/java/io/sphere/sdk/types/TypeFixtures.java [hunk 1]
```diff
@@ -34,7 +35,7 @@
     public static final String ENUM_FIELD_NAME = "enum-field-name";
     public static final String CAT_REFERENCE_FIELD_NAME = "catref";
     public static final String LOCALIZED_ENUM_FIELD_NAME = "localized-enum-field-name";
-    public static final Set<String> TYPE_IDS = new HashSet<>(asList(Category.resourceTypeId(), Customer.resourceTypeId(), Cart.resourceTypeId(), Order.resourceTypeId(), LineItem.resourceTypeId(), CustomLineItem.resourceTypeId()));
+    public static final Set<String> TYPE_IDS = new HashSet<>(asList(Category.resourceTypeId(), Customer.resourceTypeId(), Cart.resourceTypeId(), Order.resourceTypeId(), LineItem.resourceTypeId(), CustomLineItem.resourceTypeId(), Payment.resourceTypeId(), "payment-interface-interaction"));
```
Comment (cneijenhuis) [inline sphere-models/src/it/java/io/sphere/sdk/types/TypeFixtures.java]: Can we make the `payment-interface-interaction` somehow more accessible?

**Comment by cneijenhuis** [comment]
When I try to access resource on a PaymentCreatedMessage, I get a NPE.
The message:
```
{
      "id": "c7e521ad-e1ed-4dcf-bb0e-8725f2dd154e",
      "version": 1,
      "sequenceNumber": 1,
      "resource": {
        "typeId": "payment",
        "id": "68e5a12a-16e5-48d3-8e6e-bd88797f9fce"
      },
      "resourceVersion": 1,
      "type": "PaymentCreated",
      "payment": {
        "id": "68e5a12a-16e5-48d3-8e6e-bd88797f9fce",
        "version": 1,
        "customer": {
          "typeId": "customer",
          "id": "444df5fa-ffb8-415e-9755-448d3ddce32f"
        },
        "amountPlanned": {
          "currencyCode": "EUR",
          "centAmount": 2300
        },
        "paymentMethodInfo": {},
        "paymentStatus": {},
        "transactions": [],
        "interfaceInteractions": [
          {
            "type": {
              "typeId": "type",
              "id": "92be752e-2506-422e-945d-984078d1a876"
            },
            "fields": {
              "token": "tok_16vwenJxZQmu20RAmUhrbHTE"
            }
          }
        ],
        "createdAt": "2015-10-14T21:19:33.985Z",
        "lastModifiedAt": "2015-10-14T21:19:33.985Z",
        "lastMessageSequenceNumber": 1
      },
      "createdAt": "2015-10-14T21:19:34.053Z",
      "lastModifiedAt": "2015-10-14T21:19:34.053Z"
    }
```
Java-Code:
```
msg.getId(); // Works
msg.getPayment(); // Works
msg.getResource(); // Throws:
java.lang.NullPointerException
	at com.fasterxml.jackson.databind.type.TypeFactory.constructType(TypeFactory.java:455)
	at com.fasterxml.jackson.databind.ObjectMapper.readerFor(ObjectMapper.java:3236)
	at io.sphere.sdk.json.SphereJsonUtils.lambda$readObject$32(SphereJsonUtils.java:168)
	at io.sphere.sdk.json.SphereJsonUtils$$Lambda$109/398572781.get(Unknown Source)
	at io.sphere.sdk.json.SphereJsonUtils.executing(SphereJsonUtils.java:258)
	at io.sphere.sdk.json.SphereJsonUtils.readObject(SphereJsonUtils.java:168)
	at io.sphere.sdk.messages.GenericMessageImpl.getResource(GenericMessageImpl.java:63)
	at PaymentCreationListener.paymentCreated(PaymentCreationListener.java:31)
```
:question:

Happens whether I try to expand the resource or not.

**Comment by schleichardt** [comment]
@cneijenhuis can you give me the code how you query, please?

**Comment by cneijenhuis** [comment]
```
MessageQuery.of()
                        .withSort(m -> m.createdAt().sort().desc())
                        //.withExpansionPaths(m -> m.resource()) // Doesn't matter
                        .forMessageType(PaymentCreatedMessage.MESSAGE_HINT)
```

**Comment by schleichardt** [comment]
I will push a fix soon.

https://github.com/sphereio/sphere-jvm-sdk/issues/795

**Comment by schleichardt** [comment]
@cneijenhuis can you try again, please?

Sorry for the inconvenience.

**Comment by cneijenhuis** [comment]
Yes, works. Thanks!

**Pull request merged**